### ISSUE TITLE
RazorGenerator not checking IsPhysicalFileNewer

### ISSUE THREAD

**Issue by elkdanger** [RazorGenerator/RazorGenerator#78]
We currently have an issue where we always have to build our project to see changes to razor files, despite `UsePhysicalViewsIfNewer` being set to true.

I've set up a scenario where I can debug RazorGenerator.MVC and, putting a breakpoint into `IsPhysicalFileNewer()` I can see that it's not even being called by the framework after the first build for the particular view that I'm interested in.

I'm not sure where to progress to next in order to fix this - any ideas?

**Comment by davidebbo** [comment]
I just tried with my test project https://github.com/davidebbo-test/RazorGeneratorMvc51 and it works fine:

- when first running it goes against the precompiled library
- after changing home\index.cshtml and refreshing, it picks up the change

Does your issue repro with my solution? If not, then you'll have to isolate what could be different in yours.

**Comment by elkdanger** [comment]
No, your solution works fine.

I've narrowed it down to a scenario where, once the VirtualPathProviderViewEngine has cached the view location, it never seems to query whether or not the file exists until the cache is invalidated after 15 minutes, and thus never calls into `IsPhysicalFileNewer()`. Essentially once the view location is cached, the app will always use the compiled version of the view.

I have found a bug somewhere though, inside VirtualPathProviderViewEngine that PrecompiledViewEngine relies on. The issue with it not reloading the view only manifests when a general name is used for the view (as opposed to a specific virtual path), as the key name that is used when the view is put into the cache is not the same one that's used when it is read from the cache, and so the view engine always calls `FileExists()` (and thus `IsPhysicalFileNewer()`).

If there is a bug here, I realise that it's not to do with RazorGenerator persay, but regarding how it interacts with the ViewLocationCache, is it by design that a view file is never checked for its existance if its location has been cached? Have you observed this behaviour before?

**Comment by monemihir** [comment]
I just had the same issue as well. It doesn't seem to be working for me either. The only thing different between @davidebbo sample solution and mine is that I have the views in a different assembly.

Is it not supposed to work in that case?

**Comment by pranavkm** [comment]
Views in a separate assembly wouldn't work. The view engine has no idea where your app is located to cause the refresh.

**Comment by monemihir** [comment]
Yeah my thoughts exactly, but just thought I'd ask. I am being lazy and not going through the code!!

**Comment by nicholashead** [comment]
Just wanted to add that we're seeing this same/related issue-- but in our case it's causing an exception to fire; we're getting an "Object reference not set to an instance of an object." error about a partial view (`_LandingPageFloorContent.cshtml`) that actually has nothing to do with the view being rendered/requested. Very odd. Here's the stack trace:

`[NullReferenceException: Object reference not set to an instance of an object.]
   ASP._Views_Shared_PinklesResponsiveElements__LandingPageFloorContent_cshtml.Execute() in C:\_repos\Hireahelper\Hireahelper.Mvc\obj\CodeGen\Views\Views\Shared\PinklesResponsiveElements\_LandingPageFloorContent.cshtml:35
   System.Web.WebPages.WebPageBase.ExecutePageHierarchy() +252
   System.Web.Mvc.WebViewPage.ExecutePageHierarchy() +147
   System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage) +121
   System.Web.Mvc.Html.PartialExtensions.Partial(HtmlHelper htmlHelper, String partialViewName, Object model, ViewDataDictionary viewData) +136
   ASP._Views_Home_Index_cshtml.Execute() in C:\_repos\Hireahelper\Hireahelper.Mvc\obj\CodeGen\Views\Home\Index.cshtml:16
   System.Web.WebPages.WebPageBase.ExecutePageHierarchy() +252
   System.Web.Mvc.WebViewPage.ExecutePageHierarchy() +147
   System.Web.WebPages.StartPage.ExecutePageHierarchy() +88
   System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage) +106
   System.Web.Mvc.ViewResultBase.ExecuteResult(ControllerContext context) +374
   System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) +89
   System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) +833
   System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) +833
   System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) +833
   System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult) +833
   System.Web.Mvc.ControllerActionInvoker.InvokeActionResultWithFilters(ControllerContext controllerContext, IList`1 filters, ActionResult actionResult) +81
   System.Web.Mvc.Async.<>c__DisplayClass21.<BeginInvokeAction>b__1e(IAsyncResult asyncResult) +186
   System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeAction(IAsyncResult asyncResult) +38
   System.Web.Mvc.Controller.<BeginExecuteCore>b__1d(IAsyncResult asyncResult, ExecuteCoreState innerState) +29
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +67
   System.Web.Mvc.Controller.EndExecuteCore(IAsyncResult asyncResult) +53
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +36
   System.Web.Mvc.Controller.EndExecute(IAsyncResult asyncResult) +38
   System.Web.Mvc.MvcHandler.<BeginProcessRequest>b__5(IAsyncResult asyncResult, ProcessRequestState innerState) +44
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +67
   System.Web.Mvc.MvcHandler.EndProcessRequest(IAsyncResult asyncResult) +38
   System.Web.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +399
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +137`

I think perhaps RazorGenerator is getting confused/collisions on view names -- we are trying to load an `Home/Index.cshtml` view inside of an Area-- but we also have a non-area `Home/Index.cshtml` in our general Views folder.

I'm going to try and dig further on it, but if this helps anyone on the team with diagnosing, great.
